Adds L2 Documentation (#11)
bristlemouth/bm_core: docs/reference/l2.md

# L2

Within the Bristlemouth stack,

> the L2 layer is responsible for monitoring how the upper layers of the IP stack interacts with the 10BASE-T1L ethernet driver.

Reviewer: ```suggestion
the L2 layer is responsible for monitoring how the upper layers of the IP stack interact with the 10BASE-T1L ethernet driver.
```

The code itself interacts and follows the below diagram:

```{image} l2.png
:align: center
:width: 1000
:class: no-scaled-link
```

## Responsibilities

The L2 layer is the glue between the network layer of the IP stack and the ethernet driver.
It controls the flow of data to and from each of those layers.

> The following is an in depth explanation of how each of its responsibilities work:

Reviewer: ```suggestion
The following is an in-depth explanation of how each of its responsibilities work:
```

- Receive Events
  - The ethernet driver will receive a packet and pass the raw data to `bm_l2_rx`

>   - Then creates an new buffer through the ip stack abstraction and copies the raw data from the packet to this buffer,

Reviewer: ```suggestion
  - Then creates a new buffer through the IP stack abstraction and copies the raw data from the packet to this buffer,
```

  - Following,
  it then queues up this data to be handled in the `bm_l2_thread`
  - During handling,

>   if the packet is a multi-cast packet (a packet that is sent to a group of interested receivers),

Reviewer: ```suggestion
  if the packet is a multicast packet (a packet that is sent to a group of interested receivers),
```

  it is transmitted down the wire to other nodes
  - The packet is then submitted to the upper layers of the IP stack with `bm_l2_submit`,
  so that it may processed
- Transmit Events
  - Transmission messages are sent through the IP stack from upper layers,

>   which is then fed through `bm_l2_link_output`

Reviewer: ```suggestion
  and are then fed through `bm_l2_link_output`
```

  - `bm_l2_link_output` then preps the message,
  utilizing the `bm_l2_prep` function,
  and sends it to a queue which is handled in the `bm_l2_thread`
  - When processed,
  the message is then sent to the network driver to be transmitted on the line
- Power Up/Down Events

>   - External modules can call the `bm_l2_netif_set_power` API to queue up a power event to turn on/off nodes downstream
>   - To note this will also kill communication the devices

Reviewer: ```suggestion
  - External modules can call `bm_l2_netif_set_power` to queue up a power event to turn on/off nodes downstream
  - Note: this will also kill communication to the downstream devices
```

- Link Up/Down Events
  - During initialization,
  a callback is passed into the ethernet driver in order to handle when the link to a port has been dropped or opened
  - When a link is brought up or down,
  an event is queued up to be handled in the `bm_l2_thread`,
  which has the port associated with the event and whether or not the link is up or down
  - When handled,

>   the a variable is set to report to outside modules whether or not the port is up or down (utilizing `bm_l2_get_port_state`),

Reviewer: ```suggestion
  a variable is set to report to outside modules whether or not the port is up or down (utilizing `bm_l2_get_port_state`),
```

>   and a callback is invoked to report to other concerned parties about the state of the port

Author: I was able to utilize some of our discussion in the short cut ticket to write this up :) @towynlin
